<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Изучение-общей-информации" data-toc-modified-id="Изучение-общей-информации-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Изучение общей информации</a></span><ul class="toc-item"><li><span><a href="#Импорт-необходимых-библиотек" data-toc-modified-id="Импорт-необходимых-библиотек-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Импорт необходимых библиотек</a></span></li><li><span><a href="#Открытие-файлов" data-toc-modified-id="Открытие-файлов-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Открытие файлов</a></span></li><li><span><a href="#Вывод-первых-десяти-строк-таблицы" data-toc-modified-id="Вывод-первых-десяти-строк-таблицы-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Вывод первых десяти строк таблицы</a></span></li><li><span><a href="#Вывод-основной-информации-о-датафрейме" data-toc-modified-id="Вывод-основной-информации-о-датафрейме-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Вывод основной информации о датафрейме</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Удаление-пропусков" data-toc-modified-id="Удаление-пропусков-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Удаление пропусков</a></span></li><li><span><a href="#Обработка-аномальных-значений" data-toc-modified-id="Обработка-аномальных-значений-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Обработка аномальных значений</a></span></li><li><span><a href="#Удаление-пропусков-(продолжение)" data-toc-modified-id="Удаление-пропусков-(продолжение)-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Удаление пропусков (продолжение)</a></span></li><li><span><a href="#Изменение-типов-данных" data-toc-modified-id="Изменение-типов-данных-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Изменение типов данных</a></span></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Обработка дубликатов</a></span></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Категоризация данных</a></span></li></ul></li><li><span><a href="#Исследование-данных" data-toc-modified-id="Исследование-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Исследование данных</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></div>

# Исследование надежности заемщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

**Цель исследования** — сделать выводы по зависимостям между категориями и возвратом кредита в срок.

**Ход исследования:**
1. Загрузка и изучение общений информации о датафрейме
2. Предобработка данных
3. Исследование данных 
4. Вывод

## Изучение общей информации

### Импорт необходимых библиотек

In [1]:
import pandas as pd

### Открытие файлов

In [2]:
try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

### Вывод первых десяти строк таблицы

In [6]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

### Вывод основной информации о датафрейме

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


### Вывод

Согласно документации к данным, датафрейм имеет следующую информацию:

- `children` — количество детей в семье;
- `days_employed` — общий трудовой стаж в днях;
- `dob_years` — возраст клиента в годах;
- `education` — уровень образования клиента;
- `education_id` — идентификатор уровня образования;
- `family_status` — семейное положение;
- `family_status_id` — идентификатор семейного положения;
- `gender` — пол клиента;
- `income_type` — тип занятости;
-`debt` — имел ли задолженность по возврату кредитов;
- `total_income` — ежемесячный доход;
- `purpose` — цель получения кредита.

При изучении общей информации были обнаружены пропуски в данных, некорректные типы данных, дубликаты, а также аномальные значения. Необходимо произвести предобработку данных.

## Предобработка данных

### Удаление пропусков

Необходимо вывести количество пропусков в каждом столбце:

In [7]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

В двух столбцах есть пропущенные значения — `days_employed` и `total_income`. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`.

In [8]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Необходимо обработать значения в этом столбце: заменить все отрицательные значения положительными с помощью метода `abs()`.

In [9]:
data['days_employed'] = data['days_employed'].abs()

Затем для каждого типа занятости выведем медианное значение трудового стажа `days_employed` в днях:

In [10]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставим их как есть. Этот столбец не понадобится для исследования.

Выведем перечень уникальных значений столбца `children`:

In [11]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

В столбце `children` есть два аномальных значения. Удалим строки, в которых встречаются такие аномальные значения из датафрейма `data`:

In [12]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

### Удаление пропусков (продолжение)

Заполним пропуски в столбце days_employed медианными значениями по каждого типа занятости income_type:

In [13]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

Убедимся, что все пропуски заполнены:

In [14]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

Заменим вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`:

In [15]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

Обработаем неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведем их к нижнему регистру:

In [16]:
data['education'] = data['education'].str.lower()

Выведем на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалим их.

In [17]:
data.duplicated().sum()

71

In [18]:
data = data.drop_duplicates()

### Категоризация данных

 На основании диапазонов, указанных ниже, создадим в датафрейме `data` столбец `total_income_category` с категориями:

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.

Используем собственную функцию с именем `categorize_income()` и метод `apply()`:

In [19]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [20]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

Выведем перечень уникальных целей взятия кредита из столбца `purpose`:

In [21]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Создадим функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

Используем собственную функцию с именем `categorize_purpose()` и метод `apply()`:

In [22]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [23]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

## Исследование данных

Проанализируем, есть ли зависимость между количеством детей и возвратом кредита в срок.

In [25]:
selection_children = data.groupby('children')['debt'].count() # размер выборки для каждой категории
debtors_children = data.groupby('children')['debt'].sum() # количество должников для каждой категории
ratio_children = debtors_children / selection_children # доля должников в категории

children_and_debt = pd.DataFrame(data={'selection': selection_children,
                            'debtors': debtors_children,
                            'ratio': ratio_children})

display(children_and_debt)

selection  debtors     ratio
children                              
0             14091     1063  0.075438
1              4808      444  0.092346
2              2052      194  0.094542
3               330       27  0.081818
4                41        4  0.097561
5                 9        0  0.000000

**Вывод:** Доля должников в каждой категории, кроме клиентов с пятью детьми, не отличается более чем на 1-2%.Однако даже разница в 1-2% для банка с миллиардным портфелем существенна. Нужно выявить зависимость сравнив самые многочисленные категории заемщиков. 
Если принять во внимание только самые многочисленные категории заемщиков, то чем больше детей, тем больше доля должников в категории. Следовательно зависимость следующая: чем больше детей, тем меньше вероятность возврата кредита в срок.

Проанализируем, есть ли зависимость между семейным положением и возвратом кредита в срок

In [26]:
selection_family_status = data.groupby('family_status')['debt'].count() # размер выборки для каждой категории
debtors_family_status = data.groupby('family_status')['debt'].sum() # количество должников для каждой категории
ratio_family_status = debtors_family_status / selection_family_status # доля должников в категории

family_status_and_debt = pd.DataFrame(data={'selection': selection_family_status,
                            'debtors': debtors_family_status,
                            'ratio': ratio_family_status})

display(family_status_and_debt.sort_values(by='debtors', ascending=False))

selection  debtors     ratio
family_status                                      
женат / замужем            12261      927  0.075606
гражданский брак            4134      385  0.093130
Не женат / не замужем       2796      273  0.097639
в разводе                   1189       84  0.070648
вдовец / вдова               951       63  0.066246

**Вывод:** Доля должников в каждой категории не отличается более чем на 1-3%, но для данного исследования разница значимаю Для категорий "женат / замужем", "в разводе" и "вдовец / вдова" доля должников минимальна - 7.6%, 7.1% и 6.6% соответственно, следовательно, для данных категорий вероятность возврата кредита в срок больше. Для категорий "гражданский брак" и "не женат / не замужем" доля должников больше - 9.3% и 9.8% соответственно, тогда и выдача кредита таким клиентам будет более рискованна.

Проанализируем, есть ли зависимость между уровнем дохода и возвратом кредита в срок.

In [27]:
selection_total_income = data.groupby('total_income_category')['debt'].count() # размер выборки для каждой категории
debtors_total_income = data.groupby('total_income_category')['debt'].sum() # количество должников для каждой категории
ratio_total_income = debtors_total_income / selection_total_income # доля должников в категории

total_income_and_debt = pd.DataFrame(data={'selection': selection_total_income,
                            'debtors': debtors_total_income,
                            'ratio': ratio_total_income})

display(total_income_and_debt.sort_values(by='debtors', ascending=False))

selection  debtors     ratio
total_income_category                              
C                          15921     1353  0.084982
B                           5014      354  0.070602
D                            349       21  0.060172
A                             25        2  0.080000
E                             22        2  0.090909

Вывод: Если брать две самые многочисленные выборки, категория "C" - 50001–200000 и категория "B" - 200001–1000000, для которых доля должников 8.5% и 7.1% соответственно, то вероятность возврата кредита в срок будет больше у категории "B".

Проанализируем, как разные цели кредита влияют на его возврат в срок.

In [29]:
selection_purpose_category = data.groupby('purpose_category')['debt'].count()
debtors_purpose_category = data.groupby('purpose_category')['debt'].sum()
ratio_purpose_category = debtors_purpose_category / selection_purpose_category

purpose_category_and_debt = pd.DataFrame(data={'selection': selection_purpose_category,
                            'debtors': debtors_purpose_category,
                            'ratio': ratio_purpose_category})

display(purpose_category_and_debt.sort_values(by='debtors', ascending=False))

selection  debtors     ratio
purpose_category                                      
операции с недвижимостью      10751      780  0.072551
операции с автомобилем         4279      400  0.093480
получение образования          3988      369  0.092528
проведение свадьбы             2313      183  0.079118

**Вывод:** В категориях "операции с недвижимостью" и "проведение свадьбы" доли должников составляют 7.3% и 7.9% соответственно, а в категориях "операции с автомобилем" и "получение образования" - 9.3%. Тогда первые две категории будут являться менее рискованными, а последние - более рискованными.

## Вывод

Возможные причины появления пропусков в исходных данных следующие:
* Неправильный ввод данных, какие-либо ошибки при передаче или считывании данных;
* Специальные пропуски в данных с целью сокрыть какую-либо информацию.

Заполнить пропуски медианным значением — лучшее решение для количественных переменных. С практической точки зрения неверным решением будет удалять столбцы или строки с пропусками, так как последующие вычисления и выводы могут быть менее репрезентативными, поэтому пропуски нужно заполнять. Заполнять пропуски медианным значением лучше, чем средним, так как медианное значение не чувствительно к необычным максимальным или минимальным значениям.

Выводы по зависимостям между категориями и возвратом кредита в срок:

Чем больше детей у клиента, тем меньше вероятность возврата кредита в срок;
Цели кредита "операции с недвижимостью" и "проведение свадьбы" наименее рискованные;
Для категорий "женат / замужем", "в разводе" и "вдовец / вдова" вероятность возврата кредита в срок больше, чем у остальных категорий;
У уровня дохода "B" вероятность возврата кредита в срок больше, чем у категории "C", но недостаточно информации по категориям "D", "A" и "E".